File: lib/data/asset_routine_repository.dart
import 'dart:convert';

import 'package:flutter/services.dart';

import '../domain/model/class_slot.dart';
import '../domain/model/routine_day.dart';
import '../domain/model/routine_meta.dart';
import 'local_routine_store.dart';
import 'routine_file_dto.dart';

class AssetRoutineRepository {
  AssetRoutineRepository._(this.meta, this.slots, {this.hasRoutine = true});

  final RoutineMeta meta;
  final List<ClassSlot> slots;
  final bool hasRoutine;

  static const assetPath = 'assets/routine/cse_summer_2026_v5.json';
  static const pdfAsset = 'assets/routine/cse_summer_2026_v5.pdf';

  static AssetRoutineRepository empty() {
    return AssetRoutineRepository._(
      const RoutineMeta(
        department: '',
        version: '',
        semester: '',
        effectiveFrom: '',
        sourcePdf: '',
      ),
      const [],
      hasRoutine: false,
    );
  }

  static AssetRoutineRepository fromFile(RoutineFileDto file) {
    return AssetRoutineRepository._(
      RoutineMeta(
        schemaVersion: file.meta.schemaVersion,
        department: file.meta.department,
        version: file.meta.version,
        semester: file.meta.semester,
        effectiveFrom: file.meta.effectiveFrom,
        sourcePdf: file.meta.sourcePdf,
      ),
      file.slots
          .map(
            (dto) => ClassSlot(
              day: RoutineDay.fromName(dto.day),
              slot: dto.slot,
              start: dto.start,
              end: dto.end,
              course: dto.course,
              group: dto.group,
              teacher: dto.teacher,
              room: dto.room,
            ),
          )
          .toList(),
      hasRoutine: file.slots.isNotEmpty,
    );
  }

  static Future<RoutineFileDto> loadBundledFile() async {
    final raw = await rootBundle.loadString(assetPath);
    return RoutineFileDto.fromJson(jsonDecode(raw) as Map<String, dynamic>);
  }

  static Future<AssetRoutineRepository> load([LocalRoutineStore? store]) async {
    if (store != null) {
      final saved = await store.loadUser();
      if (saved == null) return AssetRoutineRepository.empty();
      if (saved.meta.origin == 'bundled') {
        return fromFile(await loadBundledFile());
      }
      return fromFile(saved);
    }
    return fromFile(await loadBundledFile());
  }
}

File: app/src/main/java/com/ababilx/routinescrapper/data/AssetRoutineRepository.kt
package com.ababilx.routinescrapper.data

import android.content.Context
import com.ababilx.routinescrapper.domain.model.ClassSlot
import com.ababilx.routinescrapper.domain.model.RoutineDay
import com.ababilx.routinescrapper.domain.model.RoutineMeta
import kotlinx.serialization.json.Json

class AssetRoutineRepository(context: Context) {
    private val json = Json { ignoreUnknownKeys = true }
    private val file: RoutineFileDto = context.assets
        .open(ASSET_PATH)
        .bufferedReader()
        .use { reader -> json.decodeFromString(reader.readText()) }

    val meta: RoutineMeta = RoutineMeta(
        department = file.meta.department,
        version = file.meta.version,
        semester = file.meta.semester,
        effectiveFrom = file.meta.effectiveFrom,
        sourcePdf = file.meta.sourcePdf,
    )

    val slots: List<ClassSlot> = file.slots.map { dto ->
        ClassSlot(
            day = RoutineDay.fromName(dto.day),
            slot = dto.slot,
            start = dto.start,
            end = dto.end,
            course = dto.course,
            group = dto.group,
            teacher = dto.teacher,
            room = dto.room,
        )
    }

    companion object {
        const val ASSET_PATH = "routine/cse_summer_2026_v5.json"
        const val PDF_ASSET = "routine/cse_summer_2026_v5.pdf"
    }
}
